File: services/sandbox-orchestrator/dynamodb_store.py
"""DynamoDB-backed sandbox registry for conversation → ECS task mapping."""

import time
import logging
from typing import Optional

import boto3
from botocore.exceptions import ClientError

logger = logging.getLogger(__name__)

# Time-to-live: 7 days from last activity
TTL_SECONDS = 7 * 24 * 3600


class SandboxRecord:
    """Represents a sandbox entry in the registry."""

    def __init__(self, data: dict):
        self.conversation_id: str = data.get('conversation_id', '')
        self.user_id: str = data.get('user_id', '')
        self.task_arn: str = data.get('task_arn', '')
        self.task_ip: str = data.get('task_ip', '')
        self.status: str = data.get('status', '')
        self.session_api_key: str = data.get('session_api_key', '')
        self.agent_server_port: int = int(data.get('agent_server_port', 8000))
        self.sandbox_spec_id: str = data.get('sandbox_spec_id', '')
        self.last_activity_at: int = int(data.get('last_activity_at', 0))
        self.created_at: int = int(data.get('created_at', 0))

    def to_dict(self) -> dict:
        return {
            'conversation_id': self.conversation_id,
            'user_id': self.user_id,
            'task_arn': self.task_arn,
            'task_ip': self.task_ip,
            'status': self.status,
            'session_api_key': self.session_api_key,
            'agent_server_port': self.agent_server_port,
            'sandbox_spec_id': self.sandbox_spec_id,
            'last_activity_at': self.last_activity_at,
            'created_at': self.created_at,
        }


class DynamoDBStore:
    """DynamoDB operations for sandbox registry."""

    def __init__(self, table_name: str, region: Optional[str] = None):
        self._table_name = table_name
        session = boto3.Session(region_name=region) if region else boto3.Session()
        self._dynamodb = session.resource('dynamodb')
        self._table = self._dynamodb.Table(table_name)

    def put_sandbox(self, record: SandboxRecord) -> None:
        """Create or update a sandbox record."""
        now = int(time.time())
        item = record.to_dict()
        if not item.get('created_at'):
            item['created_at'] = now
        item['last_activity_at'] = now
        item['ttl'] = now + TTL_SECONDS

        self._table.put_item(Item=item)
        logger.info('Put sandbox record: %s', record.conversation_id)

    def get_sandbox(self, conversation_id: str) -> Optional[SandboxRecord]:
        """Get a sandbox record by conversation_id."""
        try:
            response = self._table.get_item(
                Key={'conversation_id': conversation_id}
            )
            item = response.get('Item')
            if item:
                return SandboxRecord(item)
            return None
        except ClientError as e:
            logger.error('Failed to get sandbox %s: %s', conversation_id, e)
            return None

    def batch_get_sandboxes(self, conversation_ids: list[str]) -> list[SandboxRecord]:
        """Batch get sandbox records."""
        if not conversation_ids:
            return []

        records = []
        # DynamoDB batch_get_item limit is 100 keys
        for i in range(0, len(conversation_ids), 100):
            batch = conversation_ids[i:i + 100]
            request_items = {
                self._table_name: {
                    'Keys': [{'conversation_id': cid} for cid in batch]
                }
            }
            # Retry loop for UnprocessedKeys
            while request_items:
                response = self._dynamodb.batch_get_item(RequestItems=request_items)
                items = response.get('Responses', {}).get(self._table_name, [])
                records.extend(SandboxRecord(item) for item in items)
                request_items = response.get('UnprocessedKeys', {})

        return records

    def update_status(self, conversation_id: str, status: str,
                      task_ip: Optional[str] = None,
                      task_arn: Optional[str] = None) -> None:
        """Update sandbox status and optionally task IP/ARN."""
        now = int(time.time())
        update_expr = 'SET #status = :status, last_activity_at = :now, #ttl = :ttl'
        expr_values: dict = {
            ':status': status,
            ':now': now,
            ':ttl': now + TTL_SECONDS,
        }
        expr_names: dict = {
            '#status': 'status',
            '#ttl': 'ttl',
        }

        if task_ip is not None:
            update_expr += ', task_ip = :ip'
            expr_values[':ip'] = task_ip

        if task_arn is not None:
            update_expr += ', task_arn = :arn'
            expr_values[':arn'] = task_arn

        self._table.update_item(
            Key={'conversation_id': conversation_id},
            UpdateExpression=update_expr,
            ExpressionAttributeValues=expr_values,
            ExpressionAttributeNames=expr_names,
        )
        logger.info('Updated sandbox %s status to %s', conversation_id, status)

    def update_activity(self, conversation_id: str) -> None:
        """Update last_activity_at timestamp (called on each proxied request)."""
        now = int(time.time())
        self._table.update_item(
            Key={'conversation_id': conversation_id},
            UpdateExpression='SET last_activity_at = :now, #ttl = :ttl',
            ExpressionAttributeValues={
                ':now': now,
                ':ttl': now + TTL_SECONDS,
            },
            ExpressionAttributeNames={'#ttl': 'ttl'},
        )

    def query_by_status(self, status: str) -> list[SandboxRecord]:
        """Query sandboxes by status (via status GSI)."""
        response = self._table.query(
            IndexName='status-index',
            KeyConditionExpression='#status = :s',
            ExpressionAttributeValues={':s': status},
            ExpressionAttributeNames={'#status': 'status'},
        )
        return [SandboxRecord(item) for item in response.get('Items', [])]

    def list_running(self) -> list[SandboxRecord]:
        """List all running sandboxes (via status GSI)."""
        response = self._table.query(
            IndexName='status-index',
            KeyConditionExpression='#status = :running',
            ExpressionAttributeValues={':running': 'RUNNING'},
            ExpressionAttributeNames={'#status': 'status'},
        )
        return [SandboxRecord(item) for item in response.get('Items', [])]

    def list_by_user(self, user_id: str) -> list[SandboxRecord]:
        """List sandboxes for a specific user (via user_id GSI)."""
        response = self._table.query(
            IndexName='user_id-index',
            KeyConditionExpression='user_id = :uid',
            ExpressionAttributeValues={':uid': user_id},
        )
        return [SandboxRecord(item) for item in response.get('Items', [])]

    def query_idle(self, idle_before: int) -> list[SandboxRecord]:
        """Query running sandboxes that have been idle since before the given timestamp."""
        response = self._table.query(
            IndexName='status-index',
            KeyConditionExpression='#status = :running AND last_activity_at < :cutoff',
            ExpressionAttributeValues={
                ':running': 'RUNNING',
                ':cutoff': idle_before,
            },
            ExpressionAttributeNames={'#status': 'status'},
        )
        # Note: DynamoDB supports <, >, <=, >=, BETWEEN, begins_with on sort keys
        # in KeyConditionExpression. This is valid since last_activity_at is the sort key.
        return [SandboxRecord(item) for item in response.get('Items', [])]

File: services/sandbox-orchestrator/src/dynamodb-store.ts
/** DynamoDB-backed sandbox registry for conversation -> ECS task mapping. */

import {
  DynamoDBClient,
  BatchGetItemCommand,
  type AttributeValue,
} from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  PutCommand,
  GetCommand,
  UpdateCommand,
  QueryCommand,
} from '@aws-sdk/lib-dynamodb';
import { NodeHttpHandler } from '@smithy/node-http-handler';
import { Agent } from 'node:http';
import type { SandboxRecord, SandboxStatus } from './types.js';

/** Time-to-live: 7 days from last activity. */
const TTL_SECONDS = 7 * 24 * 3600;

export class DynamoDBStore {
  private readonly tableName: string;
  private readonly docClient: DynamoDBDocumentClient;
  private readonly rawClient: DynamoDBClient;

  constructor(tableName: string, region?: string) {
    this.tableName = tableName;
    // Persistent HTTP connections for lower latency
    this.rawClient = new DynamoDBClient({
      region,
      requestHandler: new NodeHttpHandler({
        httpAgent: new Agent({ keepAlive: true, maxSockets: 50 }),
      }),
    });
    this.docClient = DynamoDBDocumentClient.from(this.rawClient, {
      marshallOptions: { removeUndefinedValues: true },
    });
  }

  async putSandbox(record: SandboxRecord): Promise<void> {
    const now = Math.floor(Date.now() / 1000);
    const item = {
      ...record,
      created_at: record.created_at || now,
      last_activity_at: now,
      ttl: now + TTL_SECONDS,
    };
    await this.docClient.send(
      new PutCommand({ TableName: this.tableName, Item: item }),
    );
  }

  async getSandbox(conversationId: string): Promise<SandboxRecord | null> {
    const result = await this.docClient.send(
      new GetCommand({
        TableName: this.tableName,
        Key: { conversation_id: conversationId },
      }),
    );
    return (result.Item as SandboxRecord) ?? null;
  }

  async batchGetSandboxes(conversationIds: string[]): Promise<SandboxRecord[]> {
    if (!conversationIds.length) return [];

    const records: SandboxRecord[] = [];
    // DynamoDB batch_get_item limit is 100 keys
    for (let i = 0; i < conversationIds.length; i += 100) {
      const batch = conversationIds.slice(i, i + 100);
      let requestItems: Record<string, { Keys: Record<string, { S: string }>[] }> | undefined = {
        [this.tableName]: {
          Keys: batch.map((cid) => ({ conversation_id: { S: cid } })),
        },
      };

      while (requestItems && Object.keys(requestItems).length > 0) {
        const result = await this.rawClient.send(
          new BatchGetItemCommand({ RequestItems: requestItems }),
        );
        const items = result.Responses?.[this.tableName] ?? [];
        for (const item of items) {
          records.push(this.unmarshallRecord(item));
        }
        requestItems = result.UnprocessedKeys as typeof requestItems;
      }
    }
    return records;
  }

  /** Atomically claim a WARM record — returns true if this caller won the race. */
  async claimWarmTask(conversationId: string): Promise<boolean> {
    const now = Math.floor(Date.now() / 1000);
    try {
      await this.docClient.send(
        new UpdateCommand({
          TableName: this.tableName,
          Key: { conversation_id: conversationId },
          UpdateExpression: 'SET #status = :claimed, last_activity_at = :now, #ttl = :ttl',
          ConditionExpression: '#status = :warm',
          ExpressionAttributeValues: {
            ':claimed': 'CLAIMED',
            ':warm': 'WARM',
            ':now': now,
            ':ttl': now + TTL_SECONDS,
          },
          ExpressionAttributeNames: { '#status': 'status', '#ttl': 'ttl' },
        }),
      );
      return true;
    } catch (err: any) {
      if (err.name === 'ConditionalCheckFailedException') return false;
      throw err;
    }
  }

  async updateStatus(
    conversationId: string,
    status: SandboxStatus,
    taskIp?: string,
    taskArn?: string,
  ): Promise<void> {
    const now = Math.floor(Date.now() / 1000);
    let updateExpr = 'SET #status = :status, last_activity_at = :now, #ttl = :ttl';
    const exprValues: Record<string, unknown> = {
      ':status': status,
      ':now': now,
      ':ttl': now + TTL_SECONDS,
    };
    const exprNames: Record<string, string> = {
      '#status': 'status',
      '#ttl': 'ttl',
    };

    if (taskIp !== undefined) {
      updateExpr += ', task_ip = :ip';
      exprValues[':ip'] = taskIp;
    }
    if (taskArn !== undefined) {
      updateExpr += ', task_arn = :arn';
      exprValues[':arn'] = taskArn;
    }

    await this.docClient.send(
      new UpdateCommand({
        TableName: this.tableName,
        Key: { conversation_id: conversationId },
        UpdateExpression: updateExpr,
        ExpressionAttributeValues: exprValues,
        ExpressionAttributeNames: exprNames,
      }),
    );
  }

  async updateActivity(conversationId: string): Promise<void> {
    const now = Math.floor(Date.now() / 1000);
    await this.docClient.send(
      new UpdateCommand({
        TableName: this.tableName,
        Key: { conversation_id: conversationId },
        UpdateExpression: 'SET last_activity_at = :now, #ttl = :ttl',
        ExpressionAttributeValues: {
          ':now': now,
          ':ttl': now + TTL_SECONDS,
        },
        ExpressionAttributeNames: { '#ttl': 'ttl' },
      }),
    );
  }

  async queryByStatus(status: SandboxStatus): Promise<SandboxRecord[]> {
    const result = await this.docClient.send(
      new QueryCommand({
        TableName: this.tableName,
        IndexName: 'status-index',
        KeyConditionExpression: '#status = :s',
        ExpressionAttributeValues: { ':s': status },
        ExpressionAttributeNames: { '#status': 'status' },
      }),
    );
    return (result.Items as SandboxRecord[]) ?? [];
  }

  async listRunning(): Promise<SandboxRecord[]> {
    return this.queryByStatus('RUNNING');
  }

  /** Unmarshall a raw DynamoDB attribute-map into a SandboxRecord. */
  private unmarshallRecord(item: Record<string, AttributeValue>): SandboxRecord {
    return {
      conversation_id: item.conversation_id?.S ?? '',
      user_id: item.user_id?.S ?? '',
      task_arn: item.task_arn?.S ?? '',
      task_ip: item.task_ip?.S ?? '',
      status: (item.status?.S ?? '') as SandboxStatus,
      session_api_key: item.session_api_key?.S ?? '',
      agent_server_port: parseInt(item.agent_server_port?.N ?? '8000', 10),
      sandbox_spec_id: item.sandbox_spec_id?.S ?? '',
      last_activity_at: parseInt(item.last_activity_at?.N ?? '0', 10),
      created_at: parseInt(item.created_at?.N ?? '0', 10),
    };
  }
}
